Favorite nft webview network

DarkRavenn opened the pull request:
Эпик "Профиль". Часть 3. 

[decompositionProfileScreen.md](https://github.com/DarkRavenn/FakeNFT/blob/profile-module/docs/decompositionProfileScreen.md)

архитектура - MVP 
верстка - код
[доска](https://github.com/users/DarkRavenn/projects/3/views/1)  

---

### План: 
### 3. Экран "Избранные NFT" и webView ≈ **6h15m - 7h15m**
**3.1** Экран "Избранные NFT" на мок данных **≈ 2h45m - 3h45m**
  * Создать массив mockFavoriteNFTs. **30m**
  * Реализовать UICollectionView, создать кастомную ячейку с изображением, названием, рейтингом и ценой. **2-3h**
  * Добавить обработку пустого состояния. **15m**

**3.2** Экран "Избранные NFT" на данных с сервера **≈ 2h**
  * Запрос и парсинг данных через NetworkClient. **1h**
  * Обновление UI после получения данных. **1h**

**3.3** webView **≈ 1h30m**
  * Реализовать экран WebViewController. **1h**
  * Открывать "О разработчике". в WebViewController **30m**

---

### Фактически: 
### 3. Экран "Избранные NFT" и webView **9h5m**
**3.1** Экран "Избранные NFT" на мок данных **3h5m**
  * Создать массив mockFavoriteNFTs. **30m**
  * Реализовать UICollectionView, создать кастомную ячейку с изображением, названием, рейтингом и ценой. **2h20m**
  * Добавить обработку пустого состояния. **15m**

**3.2** Экран "Избранные NFT" на данных с сервера **4h**
  * Запрос и парсинг данных через NetworkClient. **1h**
  * Обновление UI после получения данных. **3h**

**3.3** webView **2h**
  * Реализовать экран WebViewController. **1h30m**
  * Открывать "О разработчике". в WebViewController **30m**

DarkRavenn committed:
Заготовка взаимодействие контроллера и презентера и добавлены мок данные в массив favoriteNFTs

DarkRavenn committed:
Экран "Избранные NFT" на мок данных

DarkRavenn committed:
Загрузка избранных нфт 12м

DarkRavenn committed:
Обработка нажатия на лайк с использованием сети

DarkRavenn committed:
обработка нажатия на лайк на экране Мои NFT

DarkRavenn committed:
WebView для отображения сайта разработчика

DarkRavenn committed:
Рефакторинг: Реорганизация кода

Helvetii951 commented on FakeNFT/Foundation/WebView/WebViewController.swift:
```diff
@@ -0,0 +1,95 @@
+import UIKit
+import WebKit
+
+final class WebViewController: UIViewController {
+    
+    // MARK: - Private Properties
+    private var urlString: String
+    private var webView: WKWebView!
```
Молодец, что реализовал контоллер с помощью веб-вью. 

Но есть вариант попроще. У нас цель - просто показать URL. Для этого очень подойдет [SFSafariViewController](https://developer.apple.com/news/?id=trjs0tcd). За 4 строчки кода мы создадим контроллер с плюшками Safari.

Helvetii951 commented on FakeNFT/Scenes /Profile/FavoriteNFT/FavoritesNFTController.swift:
```diff
@@ -20,16 +29,137 @@ final class FavoritesNFTController: UIViewController {
     
+    private lazy var collectionView: UICollectionView = {
+        let layout = UICollectionViewFlowLayout()
+        layout.itemSize = CGSize(width: (view.frame.width - 8 - 16 * 2) / 2, height: 80)
```
`- 8 - 16 * 2) / 2`. Это - [магические числа.](https://bureau.ru/soviet/20230406/). Желательно выносить такие значения в свойства или структуры с понятным названием.

Helvetii951 commented on FakeNFT/Scenes /Profile/FavoriteNFT/FavoritesNFTPresenter.swift:
```diff
@@ -0,0 +1,145 @@
+            
+            var failedNftIds: [String] = []
+            var loadedNfts: [Nft] = []
+            let dispatchGroup = DispatchGroup()
```
`DispatchGroup` 👍

Helvetii951 commented on FakeNFT/Scenes /Profile/MyNFT/MyNFTPresenter.swift:
```diff
@@ -93,9 +127,56 @@ final class MyNFTPresenter: MyNFTPresenterProtocol {
+            return true
+        } else {
+            return false
+        }
```
```
return userProfile.likes.contains(nft.id)
```

DarkRavenn committed:
Правки по ревью:

- скрыт таб бар на экране Избранные NFT
- сокращена запись условия в nftIsLiked
- исправлены магические числа в collectionView
- webView заменен на safariViewController
